KEP-4816 DRA Prioritized List design update (#5065)
kubernetes/enhancements: keps/sig-scheduling/4816-dra-prioritized-list/README.md

if it is unable to satisfy the first item, and so on. This extension to the API
is not breaking, but it makes for a somewhat awkward API where many fields on
the `DeviceRequest` (in fact all existing fields except `Name`) must not be
set whenever `FirstAvailable` is set. To avoid this API in the long term, we are
adding the `v1beta2` API described next.

For the `v1beta2` API, the `DeviceRequest` object will be restructured
so that there will be the `Name` field just like today, and for supporting
either a single request or a prioritized list of subrequests, there will
be separate fields:

* `Exactly` for a request of type `SpecificDeviceRequest` without alternatives.
`SpecificDeviceRequest` will include all the fields that exists on the
`DeviceRequest` type in `v1beta1`, except the `Name` field.
* `FirstAvailable` for providing a prioritized list of requests, each of
type `DeviceSubRequest`. The `DeviceSubRequest` type is similar to
`SpecificDeviceRequest`, except for the `AdminAccess` field that is not
available when providing multiple alternatives. The list provided in the
`FirstAvailable` field is considered a priority order, such that the

> scheduler will only try to use the second item in the list if it is unable

Reviewer: "try to use" will require a more precise definition. IIUC, a Node could now pass the Filter if any of the `FirstAvailable` can be met. And then we would have to do something in the Score phase to prefer the node that satisfies one of the earlier items.

Not saying it has to be fully explained here, but it probably has to be explained in the API comments and the documentation.

Author: I've updated the language here and also the comments in the code snippets. I've tried to make the behavior clearer and also point out the limitations here since this change does not add scoring (which is being discussed in https://github.com/kubernetes/enhancements/issues/4970)

types of devices for the same claim, and apply constraints and configuration
across those different requests.

The proposal restructures the `DeviceRequest` object so there will be a
`Name` field just like today, but for supporting either a single request

> or a prioritized list there will be the fields `Exactly` for a request without

Reviewer 2: Consider simplifying the sentences, for instance: 

"The proposal restructures the `DeviceRequest` object to support either a single request or a prioritized list. There will be a `Name` field just like today, but there will be the fields:
- `Exactly` for a request without alternatives (type `SpecificDeviceRequest`) and
- `FirstAvailable` for providing a prioritized list of requests (type `DeviceSubRequest`)

The other fields that are currently in the `DeviceRequest` object will be moved into one or both of the `SpecifiecDeviceRequest` and `DeviceSubRequest` objects."

Author: I've updated the PR based on the suggestion here.

alternatives (type `SpecificDeviceRequest`) and `FirstAvailable` for providing a
prioritized list of requests (type `DeviceSubRequest`). The other fields that are
currently in the `DeviceRequest` object will be moved into one or both of the
`SpecifiecDeviceRequest` and `DeviceSubRequest` objects. 
The list provided in the `FirstAvailable` field is
considered a priority order, such that the scheduler will only try to use
the second item in the list if it is unable to satisfy the first item,
and so on. The `Exactly` and `FirstAvailable` fields are mutually exclusive.
This is a breaking change to the `v1beta1` API, so this will be added in a
new `v1beta2` resource API.

> For the `v1beta1` API, there will be a new field called `FirstAvailable`

Reviewer 2: Consider specifying the reason for those two APIs first before describing how they differ:

"To avoid introducing a breaking API change and to ensure round-tripping between the APIs, the `v1beta1` API will remain backward compatible (only extended) and the final change will be made in `v1beta2`.

As the first step, `v1beta1` API will be extended..."

Author: I've updated the PR based on the suggestion here.

which will contain an ordered list of `DeviceSubRequest` objects. In order to
satisfy the main (containing) request, exactly one of the requests listed in

> `FirstAvailable` must be satisfied. The order listed is considered a priority

Reviewer 2: s/in `FirstAvailable`/there/

order, such that the scheduler will only try to use the second item in the list
if it is unable to satisfy the first item, and so on. This extension to the API
is not breaking, but it makes for a somewhat awkward API where many fields on
the `DeviceRequest` (in fact all existing fields except `Name`) should not be
used whenever `FirstAvailable` is used. To avoid this long term we are
adding the `v1beta2` API described above, while doing the backwards-compatible
change in the `v1beta1` API to ensure round-tripping between the APIs will work.

> These changes allows some flexibility for the user to create, say, a "gpu"

Author: Fixed.

proposal will be implemented, this is the place to discuss them.
-->

For the `v1beta2` API, all fields in the current `DeviceRequest` field except
`Name` will be moved into a new `SpecificDeviceRequest` type and the `DeviceRequest`
type will get a field called `Exactly` using this type. So for a request that doesn't
provide a prioritized list of alternatives, the user will provide the `Name` and
then set the `Exactly` field which contains the details about the request.

> The `FirstAvailable` field is mutually exclusive with the `Exactly` field, and

Reviewer 3: Other possible names:
* `FirstMatch`
* `FirstFit`
* `FirstOf`
* `OneOf`

I could imagine one day, once we have scoring, there being: `Exactly`, `FirstMatch`, `BestMatch`, etc.

Author: Yeah, so I did think about other names, but I haven't found anything that is clearly better than the naming we already have here. I think the names here are pretty good and communicates the intent pretty well.

I think I slightly prefer `FirstAvailable` over `FirstMatch`, `FirstFit` and `FirstOf` since it communicates clearly that it will pick the first one that is available. I don't think I like the `OneOf` option as it sounds more like a random entry will be chosen than the alternatives being tried in order.

> allow users to specify a prioritized list of `DeviceSubRequest` objects. The

Reviewer 2: I find `DeviceSubRequest` a bit confusing, as if it was a different request type than `SpecificDeviceRequest`. In fact they seem to be almost exactly the same except of the AdminAccess field. Does this small difference justify duplicating remaining 80% of the struct?

What about keeping the common part within struct called `DeviceSpec` and move AdminAccess outside to `DeviceRequest`? AdminAccess could be used only if `Exact DeviceSpec` was provided:

 type DeviceRequest struct {
  Name string
  Exact DeviceSpec
  FirstAvailable []DeviceSpec
  AdminAccess bool  // use with Exact only
}

Author: Yeah, something similar was discussed on the PR in https://github.com/kubernetes/kubernetes/pull/128586#discussion_r1915768098. The discussion in the thread is about the `v1beta1` API, but I think some of the arguments still apply. In particular I think it is worth keeping them separate since combining them doesn't really improve the UX and by keeping them separate we can allow them to diverge further in the future without ending up with potentially multiple fields that can't be set in certain situations.

Reviewer 2: Ack. I'm not sure in which direction the API will further evolve, so it's hard to come up with other ideas.

```go
// DeviceRequest is a request for devices required for a claim.
// This is typically a request for a single resource like a device, but can
// also ask for several identical devices. With FirstAvailable it is also
// possible to provide a prioritized list of requests.
type DeviceRequest struct {
    // Name can be used to reference this request in a pod.spec.containers[].resources.claims
    // entry and in a constraint of the claim.
    //
    // References using the name in the DeviceRequest will uniquely
    // identify a request when the Exactly field is set. When the
    // FirstAvailable field is set, a reference to the name of the
    // DeviceRequest will match whatever subrequest is chosen by the
    // scheduler.
    //
    // Must be a DNS label.
    //
    // +required
    Name string

    // Exactly specifies the details for a single request that must
    // be met exactly for the request to be satisfied.
    //
    // One of Exactly or FirstAvailable must be set.
    //
    // +optional
    // +oneOf=deviceRequestType
    Exactly *SpecificDeviceRequest

    // FirstAvailable specified a list of DeviceSubRequests, exactly

>     // one of which must be satisfied in order to satisfy this request.

Reviewer: at least one of which

Author: Fixed.

    // The scheduler will attempt to satisfy the list of subrequests in
    // order, so the second will only be attempted if the first entry

>     // couldn't be satisfied.

Reviewer: Please be more precise here.

Author: I've updated the language here.

    //
    // +optional
    // +oneOf=deviceRequestType
    // +listType=atomic
    // +featureGate=DRAPrioritizedList
    FirstAvailable []DeviceSubRequest
}

> // SpecificDeviceRequest is a request for one of more identical devices.

Reviewer 4: ```suggestion
// SpecificDeviceRequest is a request for one or more identical devices.
```

type SpecificDeviceRequest struct {
    // DeviceClassName references a specific DeviceClass, which can define
    // additional configuration and selectors to be inherited by this
    // request.
    //
    // A DeviceClassName is equired. Clients must check that it is
    // indeed set. It's absence indicates that something changed in a way that
    // is not supported by the client yet, in which case it must refuse to
    // handle the request.
    //
    // Administrators may use this to restrict which devices may get
    // requested by only installing classes with selectors for permitted
    // devices. If users are free to request anything without restrictions,
    // then administrators can create an empty DeviceClass for users
    // to reference.
    //
    // +oneOf=deviceRequestType

>     DeviceClassName string `json:"deviceClassName" protobuf:"bytes,2,name=deviceClassName"`

Author: Fixed.

    // satisfied by the scheduler to satisfy this request. It tries to
    // satisfy them in the order in which they are listed here. So if
    // there are two entries in the list, the schduler will only check
    // the second one if it determines that the first one can not be used.
    //
    // DRA does not yet implement scoring, so the scheduler will
    // select the first set of devices that satisfies all the
    // requests in the claim. And if the requirements can
    // be satisfied on more than one node, other scheduling features
    // will determine which node is chosen. This means that the set of
    // devices allocated to a claim might not be the optimal set
    // available to the cluster. Scoring will be implemented later.
    //
    // +optional
    // +oneOf=deviceRequestType
    // +listType=atomic
    // +featureGate=DRAPrioritizedList
    FirstAvailable []DeviceSubRequest
}

// SpecificDeviceRequest is a request for one or more identical devices.
type SpecificDeviceRequest struct {
    // DeviceClassName references a specific DeviceClass, which can define
    // additional configuration and selectors to be inherited by this
    // request.
    //

>     // A DeviceClassName is equired. Clients must check that it is

Reviewer 4: ```suggestion
    // A DeviceClassName is required. Clients must check that it is
```

>     //
>     // A DeviceClassName is equired. Clients must check that it is
>     // indeed set. It's absence indicates that something changed in a way that
>     // is not supported by the client yet, in which case it must refuse to
>     // handle the request.
>     //
>     // Administrators may use this to restrict which devices may get
>     // requested by only installing classes with selectors for permitted
>     // devices. If users are free to request anything without restrictions,
>     // then administrators can create an empty DeviceClass for users
>     // to reference.
>     //
>     // +oneOf=deviceRequestType

Reviewer 4: ```suggestion
    //
```

This is all not applicable anymore in v1beta2, right?

>     //
>     // +oneOf=deviceRequestType

Reviewer 4: ```suggestion
    //
    // +required
```

Not applicable in v1beta2, is it?

    // If AllocationMode is ExactCount and this field is not specified, the default is one.
    //
    // +optional
    // +oneOf=AllocationMode
    Count int64

    // AdminAccess indicates that this is a claim for administrative access
    // to the device(s). Claims with AdminAccess are expected to be used for
    // monitoring or other management services for a device.  They ignore
    // all ordinary claims to the device with respect to access modes and
    // any resource allocations.
    //
    // This is an alpha field and requires enabling the DRAAdminAccess
    // feature gate. Admin access is disabled if this field is unset or
    // set to false, otherwise it is enabled.
    //
    // +optional
    // +featureGate=DRAAdminAccess
    AdminAccess *bool
}

// DeviceSubRequest describes a request for device provided in the
// claim.spec.devices.requests[].firstAvailable array. Each
// is typically a request for a single resource like a device, but can
// also ask for several identical devices.
//
// DeviceSubRequest is similar to SpecificDeviceRequest, but doesn't expose the
// AdminAccess field as that one is only supported when requesting a
// specific device.
//

> // A DeviceClassName is equired. Clients must check that it is
> // indeed set. It's absence indicates that something changed in a way that
> // is not supported by the client yet, in which case it must refuse to
> // handle the request.

Reviewer 4: ```suggestion
```

I think we don't need to repeat that here.

type DeviceSubRequest struct {
    // Name can be used to reference this subrequest in the list of constraints
    // or the list of configurations for the claim. References must use the
    // format <main request>/<subrequest>.
    //
    // Must be a DNS label.
    //
    // +required
    Name string

    // DeviceClassName references a specific DeviceClass, which can define
    // additional configuration and selectors to be inherited by this
    // subrequest.
    //
    // A class is required. Which classes are available depends on the cluster.
    //
    // Administrators may use this to restrict which devices may get
    // requested by only installing classes with selectors for permitted
    // devices. If users are free to request anything without restrictions,
    // then administrators can create an empty DeviceClass for users
    // to reference.
    //

>     // +optional

Reviewer 4: ```suggestion
    // +required
```

    // +oneOf=deviceRequestType
    Exactly *SpecificDeviceRequest

    // FirstAvailable contains subrequests, of which exactly one will be 
    // satisfied by the scheduler to satisfy this request. It tries to
    // satisfy them in the order in which they are listed here. So if
    // there are two entries in the list, the schduler will only check
    // the second one if it determines that the first one can not be used.
    //
    // DRA does not yet implement scoring, so the scheduler will
    // select the first set of devices that satisfies all the
    // requests in the claim. And if the requirements can

>     // be satisfied on more than one node, a random node

Reviewer: It's not fully random. It depends on what other plugins say. But, arguably, it's hard to predict.

I would say something like "other scheduling features will determine which node is chosen"

Author: Updated.

    // +oneOf=deviceRequestType
    DeviceClassName string

>     // FirstAvailable contains subrequests, exactly one of which must be satisfied

Reviewer: at least one of which

Author: Fixed.

Reviewer 4: "at least" implies that also more than one could be satisfied by the allocation results. That will never be the case.

How about "FirstAvailable contains subrequests, of which exactly one must be satisfied in order to satisfy this request."?

Reviewer: There could be more than one that satisfies, but only one will be picked. Not sure how to phrase it, but the above doesn't sound right.

Reviewer 4: Next try:

"FirstAvailable contains subrequests, of which exactly one will be satisfied by the scheduler to satisfy this request. It tries to satisfy them in the order in which they are listed here."

Author: Agree with that language. I have updated the PR.

    //
    // This field may only be set in the entries of DeviceClaim.Requests.
    //
    // DRA does not yet implement scoring, so the scheduler will
    // select the first set of devices that satisfies all the
    // requests in the claim. And if the requirements can
    // be satisfied on more than one node, other scheduling features
    // will determine which node is chosen. This means that the set of
    // devices allocated to a claim might not be the optimal set
    // available to the cluster. Scoring will be implemented later.
    //
    // +optional
    // +oneOf=deviceRequestType
    // +listType=atomic
    // +featureGate=DRAPrioritizedList
    FirstAvailable []DeviceSubRequest

    ...
}

// DeviceSubRequest describes a request for device provided in the
// claim.spec.devices.requests[].firstAvailable array. Each
// is typically a request for a single resource like a device, but can
// also ask for several identical devices.
//
// DeviceSubRequest is similar to Request, but doesn't expose the AdminAccess (not
// supported) or FirstAvailable (recursion not supported) fields, as those can
// only be set on the top-level request.
//

> // A DeviceClassName is currently required. Clients must check that it is
> // indeed set. It's absence indicates that something changed in a way that
> // is not supported by the client yet, in which case it must refuse to
> // handle the request.

Reviewer 4: ```suggestion
```

type DeviceSubRequest struct {
    // Name can be used to reference this subrequest in the list of constraints
    // or the list of configurations for the claim. References must use the
    // format <main request>/<subrequest>.
    //
    // Must be a DNS label.
    //
    // +required
    Name string

    // DeviceClassName references a specific DeviceClass, which can define
    // additional configuration and selectors to be inherited by this
    // subrequest.
    //
    // A class is required. Which classes are available depends on the cluster.
    //
    // Administrators may use this to restrict which devices may get
    // requested by only installing classes with selectors for permitted
    // devices. If users are free to request anything without restrictions,
    // then administrators can create an empty DeviceClass for users
    // to reference.
    //

>     // +optional

Reviewer 4: ```suggestion
    // +required
```

    // +oneOf=deviceRequestType
    DeviceClassName string

    // FirstAvailable contains subrequests, at least one of which must be
    // satisfied in order to satisfy this request. This field may only be
    // set in the entries of DeviceClaim.Requests.
    // 
    // The scheduler will attempt to satisfy the list of subrequests in
      // order, and it will select the first one that satisfies the
    // requirements of the request. So if there are two entries in the
    // list, the schduler will only check the second one if it determines
    // that the first one can not be used.
    //
    // DRA does not yet implement scoring, so the scheduler will

>     // select the first set of devices that satisfies all the

Reviewer: This seems to describe the logic locally to a node. But it doesn't explain how a node is chosen.

If a node A satisfies option 2, and node B satisfies option 3, there is nothing in this design that prevents the scheduler to choose node B.

Author: Yeah, I see what you are referring to. I had to look into the behavior here, as the scheduler plugin for DRA does not implement the `ScorePlugin` interface. From what I can tell, the behavior in this situation is that the scheduler will choose one of the nodes that satisfy the requirements.

So for the example, you are correct that node B might be chosen in this situation, even if it seems like node A would be the better alternative. I've updated the language to reflect this.

@pohly to confirm the behavior here.

Reviewer 3: Yes, that's correct. Without scoring, this is a node-local decision. When nodes are homogenous and there are plenty of resources, this is not super useful. It's useful for heterogeneous nodes (those with different types of devices, which will be the case when used along with the *partitionable* KEP), and it is useful when resources are scarce. In other words, if you stock out on H100s, this will let you use a node with A100s.

field will reference the name of the main request for requests without
alternatives that set the `Exactly` field on the request, while it will
reference the selected subrequest on the format `<main request>/<subrequest>`
for requests that provide alternatives by setting the `FirstAvailable` field.

### Scheduler Implementation

Currently, the scheduler loops through each entry in `DeviceClaim.Requests` and
tries to satisfy each one. This would work essentially the same, except that if
the `FirstAvailable` object is set, it will attempt all alternatives for a 
request until one is found or the list is exhausted.

The current implementation in the v1beta1 API uses a missing `DeviceClassName` to make sure that
an old version of the scheduler doesn't incorrectly handle a request that
sets `FirstAvailable` which it doesn't see. If this happens it will result in an
[error](https://github.com/kubernetes/kubernetes/blob/03f134461462f86239067ec20ec17a0ba892db52/staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go#L164).
For the v1beta2 API, this will be handled by the scheduler being able to

> detect claims that hasn't set the `Exactly` object on a request.

Reviewer 4: ```suggestion
detect claims that have neither the `Exactly` nor `FirstAvailable` on a request.
```

> The current implementation will navigate a depth-first search of the devices,

Reviewer: It sounds like there we no changes to scoring, is that correct?

Author: That is correct. I've updated the language in the description and in the code comments to make this clearer.
